api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java

```diff
@@ -0,0 +1,190 @@
+			result.add(newEvent);
+			
+			if (counter == topIssuesVolume) {
+				break;
+			}
+			counter++;
+		}
+		return result;
+	}
+	
+	//find critical errors from error list
+	private static List<OOReportEvent> getCriticalErrors(ApiClient apiClient, RegressionInput input,
+			Collection<EventResult> events, DateTime deploymentStart) {
+		List<OOReportEvent> result = new ArrayList<OOReportEvent>();
+		for (EventResult event : events) {
+			if (!input.criticalExceptionTypes.contains(event.name)) {
+				continue;
+			}
+			String arcLink = getArcLink(apiClient, event.id, input, deploymentStart);
+			OOReportEvent newEvent = new OOReportEvent(event, null, arcLink);
+			result.add(newEvent);
+		}
+		return result;
+	}
+	
+	//find resurfaced errors from error list
+	private static List<OOReportEvent> getResurfacedErrors(ApiClient apiClient, RegressionInput input,
+			Collection<EventResult> events, DateTime deploymentStart) {
+		List<OOReportEvent> result = new ArrayList<OOReportEvent>();
+		for (EventResult event : events) {
+			if (!event.labels.contains(RegressionStringUtil.RESURFACED_ISSUE)) {
+				continue;
+			}
+			String arcLink = getArcLink(apiClient, event.id, input, deploymentStart);
+			OOReportEvent newEvent = new OOReportEvent(event, null, arcLink);
+			result.add(newEvent);
+		}
+		return result;
+	}
+	
+	//find new errors from error list
+	private static List<OOReportEvent> getNewErrors (ApiClient apiClient, RegressionInput input, Collection<EventResult> events, DateTime deploymentStart) {		
+		List<OOReportEvent> result = new ArrayList<OOReportEvent>();		
+		for (EventResult event : events) {			
+			if (!input.deployments.contains(event.introduced_by)) {
+				continue;
+			}			
+			String arcLink = getArcLink(apiClient, event.id, input, deploymentStart);
+			OOReportEvent newEvent = new OOReportEvent(event, RegressionStringUtil.NEW_ISSUE, arcLink);
+			result.add(newEvent);
+		}
+		return result;
+	}
+	
+	//get the arclink for the event
+	public static String getArcLink(ApiClient apiClient, String eventId,
+			RegressionInput input, DateTime activeWindowStart) {
+
+		DateTime activeWndowStart = activeWindowStart;
+		DateTime from = activeWndowStart.minusMinutes(input.baselineTimespan); 
+		
+		String result = EventUtil.getEventRecentLinkDefault(apiClient, input.serviceId, eventId, 
+				from, DateTime.now(), input.applictations, input.servers, input.deployments, 
+				EventUtil.DEFAULT_PERIOD);
+
+		return result;
+	}
+	
+	//filter in or out events based on the pattern 
+	private static Collection<EventResult> filterEvents(Collection<EventResult> events, String regexFilter) {
+		if (regexFilter == null) {
+			return events;
+		}
+		
+		List<EventResult> returnEvents = new ArrayList<EventResult>();	
+		for (Iterator iterator = events.iterator(); iterator.hasNext();) {
+			EventResult event = (EventResult) iterator.next();
+			if (evaluateEvent(event, getPattern(regexFilter))) {
```

Comment:
Probably not too significant, but as `getPattern` internally calls `Pattern.compile`, it's better to extract the result `Pattern` object to a local variable outside of this loop, to avoid some overhead.

Reply:
Again, copied code. Happy to change if needed.